File: nearly-sorted-data/SelectionSort/SelectionSort.py
import time

def selection_sort(arr):
    for i in range(len(arr)):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

def main():
    try:
        with open("nearly_sorted_data.txt", "r") as file:
            arr = list(map(int, file.read().split()))
    except FileNotFoundError:
        print("Error: nearly_sorted_data.txt file not found!")
        return

    start_time = time.time()
    selection_sort(arr)
    end_time = time.time()

    execution_time = end_time - start_time
    print(f"Selection Sort Execution Time: {execution_time:.6f} seconds")

if __name__ == "__main__":
    main()

File: nearly-sorted-data/SelectionSort/SelectionSort.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <chrono>

using namespace std;

void selectionSort(vector<int> &arr) {
    int n = arr.size();

    for (int i = 0; i < n - 1; i++) {
        int minIndex = i;

        for (int j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }

        swap(arr[minIndex], arr[i]);
    }
}

int main() {
    ifstream inputFile("nearly_sorted_data.txt");
    vector<int> arr;
    int num;

    while (inputFile >> num) {
        arr.push_back(num);
    }
    inputFile.close();

    auto start = chrono::high_resolution_clock::now();
    selectionSort(arr);
    auto end = chrono::high_resolution_clock::now();

    chrono::duration<double> elapsed = end - start;
    cout << "Selection Sort Execution Time: " << elapsed.count() << " seconds" << endl;

    return 0;
}
